### ISSUE TITLE
Shape Mismatch: flattened_pred shape (505728,) vs flattened_truth shape (384,)

### ISSUE THREAD

**Issue by FatmaAbdel** [FatmaAbdel/DDI-side-effects#1]
I'm currently working on a multi-label polypharmacy side effects link prediction task using a heterogeneous network with 2 node types and 3 edge types. I have successfully preprocessed the data and split it into train, validation, and test sets. However, I’ve encountered issues while trying to evaluate the model's performance : 
1. Despite experimenting with **different values for hidden_channels and learning rates**, I consistently get **unusual results** where **Precision, Recall, F1, and Accuracy are all equal**.
2. When attempting to **display the confusion matrix, I encounter a shape mismatch error**

```
class GNN(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        self.feature_transform_drug = nn.Linear(7, hidden_channels)
        self.feature_transform_gene = nn.Linear(7, hidden_channels)

        self.conv1 = HeteroConv({
            ('drug', 'interacts_with', 'drug'): SAGEConv((hidden_channels, hidden_channels), hidden_channels),
            ('drug', 'interacts_with', 'gene'): SAGEConv((hidden_channels, hidden_channels), hidden_channels),
            ('gene', 'interacts_with', 'gene'): GATConv((hidden_channels, hidden_channels), hidden_channels, heads=1)
        }, aggr='sum')

        self.conv2 = HeteroConv({
            ('drug', 'interacts_with', 'drug'): SAGEConv((hidden_channels, hidden_channels), hidden_channels),
            ('drug', 'interacts_with', 'gene'): SAGEConv((hidden_channels, hidden_channels), hidden_channels),
            ('gene', 'interacts_with', 'gene'): GATConv((hidden_channels, hidden_channels), hidden_channels, heads=1)
        }, aggr='sum')

    def forward(self, x_dict, edge_index_dict):
        x_dict['drug'] = F.relu(self.feature_transform_drug(x_dict['drug']))
        x_dict['gene'] = F.relu(self.feature_transform_gene(x_dict['gene']))
        x_dict = self.conv1(x_dict, edge_index_dict)
        x_dict = {key: F.relu(x) for key, x in x_dict.items()}
        x_dict = self.conv2(x_dict, edge_index_dict)
        return x_dict

class MultiLabelClassifier(torch.nn.Module):
    def __init__(self, hidden_channels, num_labels):
        super().__init__()
        self.fc_input_transform = nn.Linear(hidden_channels, hidden_channels)
        self.fc = nn.Linear(hidden_channels, num_labels)  # Ensure num_labels matches the number of classes

    def forward(self, x_drug, edge_label_index):
        edge_feat_drug = x_drug[edge_label_index[0]]
        edge_feat_drug_target = x_drug[edge_label_index[1]]
        edge_feat = edge_feat_drug * edge_feat_drug_target
        edge_feat = F.relu(self.fc_input_transform(edge_feat))
        output = self.fc(edge_feat)
        return output  # Output size should now be (batch_size, num_labels)


class Model(torch.nn.Module):
    def __init__(self, hidden_channels, num_labels):
        super().__init__()
        self.gnn = GNN(hidden_channels)
        self.classifier = MultiLabelClassifier(hidden_channels, num_labels)

    def forward(self, data):
        x_dict = self.gnn(data.x_dict, data.edge_index_dict)
        pred = self.classifier(x_dict['drug'], data['drug', 'interacts_with', 'drug'].edge_label_index)
        return pred
```
Training loop : 
```
num_labels = df2['Polypharmacy Side Effect_encoded'].nunique()
model = Model(hidden_channels=64, num_labels=num_labels).to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.00001)

# Calculate class weights based on the distribution of the labels
labels = torch.cat([train_data['drug', 'interacts_with', 'drug'].edge_label,
                    val_data['drug', 'interacts_with', 'drug'].edge_label,
                    test_data['drug', 'interacts_with', 'drug'].edge_label], dim=0)

# Handle missing classes in class_counts
class_counts = torch.bincount(labels.flatten().to(torch.int64), minlength=num_labels)
total_samples = class_counts.sum().float()
class_weights = total_samples / (class_counts.float() + 1e-5)
class_weights = class_weights / class_weights.sum()

criterion = nn.BCEWithLogitsLoss(pos_weight=class_weights)

for epoch in range(1, 6):
    model.train()
    total_loss = 0
    total_examples = 0

    for sampled_data in tqdm.tqdm(train_loader):
        optimizer.zero_grad()
        sampled_data = sampled_data.to(device)

        # Forward pass
        pred = model(sampled_data)  # Size: (batch_size, num_labels)

        # Ground truth: size should be (batch_size,)
        ground_truth = sampled_data["drug", "interacts_with", "drug"].edge_label.to(torch.int64).to(device)

        # Ensure the ground truth tensor matches the number of possible labels
        one_hot_ground_truth = torch.zeros_like(pred)
        one_hot_ground_truth.scatter_(1, ground_truth.unsqueeze(1), 1)

        # Ensure dimensions match before loss calculation
        assert pred.shape == one_hot_ground_truth.shape, f"Shape mismatch: pred {pred.shape}, one_hot_ground_truth {one_hot_ground_truth.shape}"

        # Calculate loss with weighted classes
        loss = criterion(pred, one_hot_ground_truth)

        # Regularization for potential overfitting
        l2_lambda = 1e-5
        l2_norm = sum(p.pow(2.0).sum() for p in model.parameters())
        loss = loss + l2_lambda * l2_norm

        loss.backward()
        optimizer.step()

        total_loss += loss.item() * pred.size(0)
        total_examples += pred.size(0)

    print(f"Epoch: {epoch:03d}, Loss: {total_loss / total_examples:.4f}")
```
![Capture1](https://github.com/user-attachments/assets/42367a0f-02c2-462b-9dd2-a11acd35cdb7)
Performance evaluation : 
```
def evaluate(loader, model, device):
    model.eval()
    preds, ground_truths = [], []

    with torch.no_grad():
        for sampled_data in tqdm.tqdm(loader):
            sampled_data = sampled_data.to(device)
            pred = model(sampled_data)  # Size: (batch_size, num_labels)

            # Apply sigmoid to convert logits to probabilities
            pred_probs = torch.sigmoid(pred).cpu().numpy()

            # Ground truth: size should be (batch_size,)
            ground_truth = sampled_data["drug", "interacts_with", "drug"].edge_label.to(torch.int64).cpu().numpy()

            # Convert ground_truth to one-hot encoding
            one_hot_ground_truth = np.zeros_like(pred_probs)
            one_hot_ground_truth[np.arange(len(ground_truth)), ground_truth] = 1

            preds.append(pred_probs)
            ground_truths.append(one_hot_ground_truth)

    preds = np.concatenate(preds, axis=0)
    ground_truths = np.concatenate(ground_truths, axis=0)

    # Calculate metrics
    auc = roc_auc_score(ground_truths, preds, average='micro')
    precision = precision_score(ground_truths.argmax(axis=1), preds.argmax(axis=1), average='micro')
    recall = recall_score(ground_truths.argmax(axis=1), preds.argmax(axis=1), average='micro')
    f1 = f1_score(ground_truths.argmax(axis=1), preds.argmax(axis=1), average='micro')
    accuracy = accuracy_score(ground_truths.argmax(axis=1), preds.argmax(axis=1))

    return auc, precision, recall, f1, accuracy

# Validation
val_metrics = evaluate(val_loader, model, device)
print(f"Validation AUC: {val_metrics[0]:.4f}, Precision: {val_metrics[1]:.4f}, Recall: {val_metrics[2]:.4f}, F1: {val_metrics[3]:.4f}, Accuracy: {val_metrics[4]:.4f}")

# Test
test_metrics = evaluate(test_loader, model, device)
print(f"Test AUC: {test_metrics[0]:.4f}, Precision: {test_metrics[1]:.4f}, Recall: {test_metrics[2]:.4f}, F1: {test_metrics[3]:.4f}, Accuracy: {test_metrics[4]:.4f}")
```
![Capture2](https://github.com/user-attachments/assets/a9098477-faf0-420c-9bb5-630a1a420c60)

Note that the same drug pair can have 0 or multiple side effects, leading to a multi-label classification scenario, and I applied class weights to address the imbalance in the classes, but it didn't work. I’m unsure of the root cause of these issues and would appreciate any insights or suggestions
